from_variants_impl/src/from_impl.rs

```diff
@@ -144,7 +143,7 @@ mod tests {
             fi,
             quote!(
                 #[doc = "Convert into [`Bar`](#variant.Bar) variant."]
-                impl<INTO: ::from_variants::export::Into<String> >
+                impl<INTO: ::from_variants::export::Into<String>>
```

Comment:
I had to put spaces in here to avoid `>>` being interpreted as one token, breaking the output completely. Has that now been fixed?

Reply:
Was that ever a bug? I saw that `rustfmt::skip` and wondered why it was there and just saw how it only reformats the `>>`.
I don't think such bug exists, Rust's syntax clearly doesn't allow `>>` operator in type context, the lexer is should never have a problem figuring this out. It may be a problem in C++ templates where type/value context can be confused by the compiler, but such problem doesn't exist in Rust.

Reply:
I need to try this when I get home. I believe the issue is that quasi-quoting doesn't run in type context, so if it infers that you're trying to push the "shift-left" symbol that's the Token emitted, and that is incorrect.

Reply:
Hmm, that would be a pretty disruptive bug. I am pretty sure I've been spelling types like `Option<Vec<_>>` with `syn::parse_quote!()` and `quote::quote!()` and never had this problem. I've just rerun the `cargo test` and it worked for me.

I think rustc and `quote/syn` are smart enough to take apart `shift-right` operator from an angle bracket. IIRC, `rustc` only verifies that all `([{` have a matching and balanced  `}])`, and maybe several more simple invariants. 

According to the Rust docs of `proc-macro` crate for [`Punct` type](https://doc.rust-lang.org/stable/proc_macro/struct.Punct.html):
> Multi-character operators like += are represented as two instances of Punct with different forms of Spacing returned.

So, the compiler itself doesn't decide what `>>/<<` mean in the macro callsite. It passes that to the proc macro the following way. I tested it with this:

```rust
fn main() {
    proc_macro_joint::debug_log_tokens!(
        1 >> 2
    );

    proc_macro_joint::debug_log_tokens!(
        <<Option<Vec<String>>>>::None
    );
}
```

```rust
#[proc_macro]
pub fn debug_log_tokens(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    panic!("Input: {input:#?}");
}
```

The compile output is the following

<details>
<summary>Compile output</summary>

```
error: proc macro panicked
 --> crates/proc-macro-joint-main/src/main.rs:5:5
  |
5 | /     proc_macro_joint::debug_log_tokens!(
6 | |         1 >> 2
7 | |     );
  | |_____^
  |
  = help: message: Input: TokenStream [
              Literal {
                  kind: Integer,
                  symbol: "1",
                  suffix: None,
                  span: #0 bytes(64..65),
              },
              Punct {
                  ch: '>',
                  spacing: Joint,
                  span: #0 bytes(66..68),
              },
              Punct {
                  ch: '>',
                  spacing: Alone,
                  span: #0 bytes(66..68),
              },
              Literal {
                  kind: Integer,
                  symbol: "2",
                  suffix: None,
                  span: #0 bytes(69..70),
              },
          ]

error: proc macro panicked
  --> crates/proc-macro-joint-main/src/main.rs:9:5
   |
9  | /     proc_macro_joint::debug_log_tokens!(
10 | |         <<Option<Vec<String>>>>::None
11 | |     );
   | |_____^
   |
   = help: message: Input: TokenStream [
               Punct {
                   ch: '<',
                   spacing: Joint,
                   span: #0 bytes(128..130),
               },
               Punct {
                   ch: '<',
                   spacing: Alone,
                   span: #0 bytes(128..130),
               },
               Ident {
                   ident: "Option",
                   span: #0 bytes(130..136),
               },
               Punct {
                   ch: '<',
                   spacing: Alone,
                   span: #0 bytes(136..137),
               },
               Ident {
                   ident: "Vec",
                   span: #0 bytes(137..140),
               },
               Punct {
                   ch: '<',
                   spacing: Alone,
                   span: #0 bytes(140..141),
               },
               Ident {
                   ident: "String",
                   span: #0 bytes(141..147),
               },
               Punct {
                   ch: '>',
                   spacing: Joint,
                   span: #0 bytes(147..149),
               },
               Punct {
                   ch: '>',
                   spacing: Joint,
                   span: #0 bytes(147..149),
               },
               Punct {
                   ch: '>',
                   spacing: Joint,
                   span: #0 bytes(149..151),
               },
               Punct {
                   ch: '>',
                   spacing: Joint,
                   span: #0 bytes(149..151),
               },
               Punct {
                   ch: ':',
                   spacing: Joint,
                   span: #0 bytes(151..153),
               },
               Punct {
                   ch: ':',
                   spacing: Alone,
                   span: #0 bytes(151..153),
               },
               Ident {
                   ident: "None",
                   span: #0 bytes(153..157),
               },
           ]
````
</details>

This proves that `rustc` doesn't do anything fancy to analyze the semantics of angle brackets, it just leaves this for the proc macro implementation.

Then `syn::parse_quote!()` is intelligent enough to understand what `>>` means, because it has the context of the expected syn AST node type, so knows what to expect.
And regarding `quote::quote!()`: it is very dumb, it doesn't assign any semantic meaning to `>>`, it just converts each input token to `proc_macro::TokenTree` using its Spacing API. For example, for multi-symbol tokens it [does this](https://github.com/dtolnay/quote/blob/278057d0d2e59c7de95d79d42e4934783d013e83/src/runtime.rs#L306-L336). And once `quote!()` forwards the raw `TokenTree` with the spacing properly set up, `rustc` analyzes the output of the proc macro as regular Rust code, so it compile just fine

Reply:
Sorry for verbosity. I always try to fight the "I don't know how, but it works, and we should better not touch it ever"

Reply:
Thanks for putting the effort into verifying this isn't an issue! I agree with your philosophy of questioning these sorts of limitations; I've now re-enabled the CI workflow, so hopefully one more push will be all that's needed to get the tests to run and then we'll be good to go.

Really appreciate your diligence in running this down.